src/net.cpp

```diff
@@ -718,6 +720,127 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
+    }
+}
+
+CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time)
+{
+    // In v2, vRecv contains the encrypted payload plus the MAC tag (1+bytes serialized message command + ? bytes message payload + 16 byte mac tag)
+    assert(Complete());
+
+    // defensive decoding (MAC check, decryption, command deserialization)
+    // we'll always return a CNetMessage (even if encryption fails), we always increase the AEAD sequence numbers
+    bool valid_checksum = false;
+    bool valid_header = false;
+    std::string command_name;
+
+    if (m_aead->Crypt(m_payload_seqnr, m_aad_seqnr, m_aad_pos, (unsigned char*)vRecv.data(), vRecv.size(), (const uint8_t*)vRecv.data(), vRecv.size(), false)) {
+        // MAC check was successful
+        valid_checksum = true;
+
+        // okay, we could decrypt it, now remove packet length and MAC tag
+        assert(vRecv.size() > CHACHA20_POLY1305_AEAD_AAD_LEN + CHACHA20_POLY1305_AEAD_TAG_LEN);
+        // CDataStream::erase at the begin will just increase the read pos
+        vRecv.erase(vRecv.begin(), vRecv.begin() + CHACHA20_POLY1305_AEAD_AAD_LEN);
+        vRecv.erase(vRecv.end() - CHACHA20_POLY1305_AEAD_TAG_LEN, vRecv.end());
+
+        uint8_t size_or_shortid = 0;
+        try {
+            vRecv >> size_or_shortid;
+        } catch (const std::ios_base::failure&) {
+            LogPrint(BCLog::NET, "Invalid command name\n");
+        }
+        if (size_or_shortid > 0 && size_or_shortid <= 12 && vRecv.size() >= size_or_shortid) {
+            // string command
+            valid_header = true;
+
+            // use direct read since we already read the varlens size uint8_t
+            command_name.resize(size_or_shortid);
+            vRecv.read(&command_name[0], size_or_shortid);
+        }
+        // try for short ID
+        if (!valid_header && size_or_shortid >= 0) {
+            valid_header = true;
+            if (!GetCommandFromShortCommandID(size_or_shortid, command_name)) {
+                // unknown-short-id
+                //  results in a valid but unknown message (will be skipped)
+                command_name = "unknown-" + ToString(size_or_shortid);
+            }
+        }
+    }
+    // increase the aad_pos and eventually the sequence number
+    m_aad_pos += CHACHA20_POLY1305_AEAD_AAD_LEN;
+    if (m_aad_pos + CHACHA20_POLY1305_AEAD_AAD_LEN > CHACHA20_ROUND_OUTPUT) {
+        m_aad_pos = 0;
+        m_aad_seqnr++;
+    }
+    // increase the payload sequence number by 1
+    m_payload_seqnr++;
+
+    // decompose a single CNetMessage from the TransportDeserializer
+    CNetMessage msg(std::move(vRecv)); // result in a message with CDataStream with readpos pointing to the message payload
+    msg.m_command = command_name;
+
+    // store state about valid header, netmagic and checksum
+    msg.m_valid_header = valid_header; // not relevant for v2, always pass
+    msg.m_valid_checksum = valid_checksum;
```

Comment:
I think checksum is also not revelant and always true, so comment may be generic for three of them? But generally if this field are always true for v2, a future refactor may just call public methods on messages in `ProcessMessage` like `{Header,Checksum,Netmagic}Valid`.

Reply:
`msg.m_valid_checksum` is a field we should populate to respect the abstract `TransportDeserializer` class.
In the v2 case, the `checksum` is the MAC. We set the field whenever the MAC has validated. How to deal with it is a matter of the layer above.